ext/mysql2/result.c

```diff
@@ -755,6 +755,134 @@ static VALUE rb_mysql_result_fetch_fields(VALUE self) {
   return wrapper->fields;
 }
 
+static void rb_mysql_row_query_options(VALUE opts, ID *db_timezone, ID *app_timezone, int *symbolizeKeys, int *asArray, int *castBool, int *cast, int *cacheRows) {
+  ID dbTz, appTz;
+
+   *symbolizeKeys = RTEST(rb_hash_aref(opts, sym_symbolize_keys));
+   *asArray       = rb_hash_aref(opts, sym_as) == sym_array;
+   *castBool      = RTEST(rb_hash_aref(opts, sym_cast_booleans));
+   *cacheRows     = RTEST(rb_hash_aref(opts, sym_cache_rows));
+   *cast          = RTEST(rb_hash_aref(opts, sym_cast));
+
+  dbTz = rb_hash_aref(opts, sym_database_timezone);
+  if (dbTz == sym_local) {
+    *db_timezone = intern_local;
+  } else if (dbTz == sym_utc) {
+    *db_timezone = intern_utc;
+  } else {
+    if (!NIL_P(dbTz)) {
+      rb_warn(":database_timezone option must be :utc or :local - defaulting to :local");
+    }
+    *db_timezone = intern_local;
+  }
+
+  appTz = rb_hash_aref(opts, sym_application_timezone);
+  if (appTz == sym_local) {
+    *app_timezone = intern_local;
+  } else if (appTz == sym_utc) {
+    *app_timezone = intern_utc;
+  } else {
+    *app_timezone = Qnil;
+  }
+}
+
+static VALUE rb_mysql_result_element(int argc, VALUE * argv, VALUE self) {
+  result_each_args args;
+  MYSQL_FIELD *fields = NULL;
+  ID db_timezone, app_timezone;
+  VALUE seek, count, row, rows;
+  long i, c_seek, c_count = 0;
+  int symbolizeKeys, asArray, castBool, cacheRows, cast;
+  VALUE defaults, block, opts, (*fetch_row_func)(VALUE, MYSQL_FIELD *fields, const result_each_args *args);
+
+  GET_RESULT(self);
+
+  defaults = rb_iv_get(self, "@query_options");
+  Check_Type(defaults, T_HASH);
+  rb_scan_args(argc, argv, "12&", &seek, &count, &opts, &block);
+
+  /* If the second arg is a hash, it's the opts and there's no count */
+  if (TYPE(count) == T_HASH) {
+    opts = count;
+    count = Qnil;
+  }
+
+  c_seek = NUM2LONG(seek);
+  if (!NIL_P(count)) {
+    c_count = NUM2LONG(count);
+    /* Special case: ary[x, 0] returns []*/
+    if (!c_count) return rb_ary_new();
+  }
+
+  if (!NIL_P(opts)) {
+    opts = rb_funcall(defaults, intern_merge, 1, opts);
```

Comment:
this preamble could be implemented in ruby and would be much easier to read - why not do that?